Add an all-in-one histogram intrinsic, along with lowering for AArch64

huntergr-arm opened the pull request:
Based on discussion from https://discourse.llvm.org/t/rfc-vectorization-support-for-histogram-count-operations/74788

Current interface is:

llvm.experimental.histogram(<vecty> ptrs, <intty> inc_amount, <vecty> mask)

The integer type used by 'inc_amount' needs to match the type of the buckets
in memory.

The intrinsic covers the following operations:
  * Gather load
  * histogram on the elements of 'ptrs'
  * multiply the histogram results by 'inc_amount'
  * add the result of the multiply to the values loaded by the gather
  * scatter store the results of the add

At this stage, we'd mostly like to discuss which type of intrinsic would be preferable (this one, or the original proposed in the RFC). There will be work on the LoopVectorize side to recognize histogram operations proceeding independently of this patch, at least for now.

huntergr-arm committed:
Add an all-in-one histogram intrinsic, along with lowering for AArch64

Current interface is:

llvm.experimental.vector.histogram.op(<vecty> ptrs, <intty> inc_amount, <vecty> mask)

Where op is the update operation (currently limited to 'add').

The integer type used by 'inc_amount' needs to match the type of the buckets
in memory.

The intrinsic covers the following operations:
  * Gather load
  * histogram on the elements of 'ptrs'
  * multiply the histogram results by 'inc_amount'
  * add the result of the multiply to the values loaded by the gather
  * scatter store the results of the add

These operations can obviously be scalarized on platforms without the
relevant instructions.

RonDahan101 commented:
This is a great improvement! I would be glad if you could consider supporting these two cases as well:
* Histograms with saturated values.
* More operations instead of only 'add' (Min/Max/Set).

SamTebbs33 reviewed (approved):
This looks good to me as is and I support the all-in-one approach. I think it would be worth addressing the FIXMEs that are simple enough to do in this patch (moving out code to their own functions and the operand assertion in `getMaskedHistogram`).

arsenm commented on llvm/include/llvm/IR/Intrinsics.td:
```diff
@@ -1848,6 +1848,13 @@ def int_experimental_vp_strided_load  : DefaultAttrsIntrinsic<[llvm_anyvector_ty
+                             [ llvm_anyvector_ty, // Vector of pointers
+                               llvm_anyint_ty,    // Increment
+                               LLVMScalarOrSameVectorWidth<0, llvm_i1_ty>], // Mask
+                             [ IntrNoSync, IntrWillReturn ]>;
```
IntrNoSync and IntrWillReturn are implied by DefaultAttrsIntrinsic

paschalis-mpeis reviewed (approved):
Looks good. Agree with proceeding with some of the FIXMEs.

huntergr-arm commented:
Rebased, addressed the easy FIXMEs.

I made the intrinsic form extensible, so that you specify the update operation instead of just assuming it's an addition. Have added to the LangRef.

huntergr-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -6673,6 +6677,56 @@ SDValue AArch64TargetLowering::LowerOperation(SDValue Op,
+    ISD::MemIndexType IndexType = HG->getIndexType();
+    SDValue Gather =
+                DAG.getMaskedGather(DAG.getVTList(MemVT, MVT::Other), MemVT, DL,
+                                    Ops, HG->getMemOperand(),
```
MemSDNodes have a member variable of a pointer to a MachineMemOperand, so to correctly have separate MOLoad semantics for the gather and MOStore semantics for the scatter we need two different MMO instances.

(I'm not sure if it matters if we just let both be `MOLoad | MOStore` at this point, I'm just being a bit cautious)

huntergr-arm replied:
Ah, I think I see what you meant now; the diff wasn't quite clear after the forced push. Done.

jayfoad commented on llvm/docs/LangRef.rst:
```diff
@@ -19030,6 +19030,65 @@ will be on any later loop iteration.
+   the ``ptrs`` operand.
+
+The ``mask`` operand will apply to at least the gather and scatter operations,
+and potentially the update if supported.
```
"potentially the update if supported" sounds like an implementation optimization. I have a slight preference to keep the semantics section as simple as possible by omitting it. Same for the "or a set of index value..." part above.

jayfoad commented on llvm/docs/LangRef.rst:
```diff
@@ -19030,6 +19030,65 @@ will be on any later loop iteration.
+updated. The second argument is a scalar used to update the value from
+memory; it must match the type of value to be updated. The final argument
+is a mask value which will exclude that vector element from being updated,
+and will exclude it from any cross-lane calculations to determine the
+final values for each memory location.
```
I'm not sure that the "and will exclude it from ..." part really adds anything useful to the description?

huntergr-arm commented:
Rebased, addressed comments. I've added scalarization of the intrinsic for targets which don't have an appropriate vector instruction to map this to.

mgabka commented on llvm/docs/LangRef.rst:
```diff
@@ -19068,6 +19068,60 @@ will be on any later loop iteration.
 This intrinsic will only return 0 if the input count is also 0. A non-zero input
 count will produce a non-zero result.
 
+'``llvm.experimental.vector.histogram.*``' Intrinsics
```
nit: looks like a singular form Intrinsics is used every where else.

RKSimon commented:
@huntergr-arm This is failing on EXPENSIVE_CHECKS builds - please can you take a look?
https://lab.llvm.org/buildbot/#/builders/16/builds/65450

huntergr-arm commented:
Hi @RKSimon -- I already have a patch for that: https://github.com/llvm/llvm-project/pull/92095
I should have mentioned it here though.
